### PR TITLE
Feat/remix recipe

### PR DESCRIPTION
**Pull request by EarthlingDavey**
Add remix recipe - I've only touched the `remix/recipe` folder. So I don't know if there's extra stuff I need to do to prepare the package for release.

It's working fine - but with some TS errors like : https://github.com/measuredco/puck/issues/185

Do you mind reviewing, and advising the best way to fix the errors.

I can also remove or amend the devcontainer if needed.

Thanks

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "3000": {
+      "label": "Hello Remote World",
+      "onAutoForward": "notify"
+    }
+  },
+  // Use 'postCreateCommand' to run commands after the container is created.
+  "postCreateCommand": "yarn install"
+  // Uncomment to connect as root instead. More info: https://aka.ms/dev-containers-non-root.
+  // "remoteUser": "root"
+}
```
Comment (chrisvxd) [inline .devcontainer/devcontainer.json:33]: We're not currently using dev containers, so I'd rather not pollute the codebase for the time-being.

Could you drop this change?

File: recipes/remix/remix.config.js [hunk 2]
```diff
@@ -5,4 +5,8 @@ export default {
   // assetsBuildDirectory: "public/build",
   // publicPath: "/build/",
   // serverBuildPath: "build/index.js",
+  // crypto is used by packages/core/dist/index.js
+  // is it possible to not send it to the browser?
+  // for not, we need to polyfill it...
+  browserNodeBuiltinsPolyfill: { modules: { crypto: true } },
```
Comment (chrisvxd) [inline recipes/remix/remix.config.js:11]: This is a bit awkward. We can hopefully remove this when we deal with #112

File: recipes/remix/app/root.tsx [hunk 3]
```diff
@@ -7,13 +7,21 @@ import {
...
 ];
 
 export default function App() {
+  /**
+   * Disable client-side JS - Optional
+   * @see https://remix.run/docs/en/main/guides/disabling-javascript
+   */
+  // const matches = useMatches();
+  // const includeScripts = matches.some((match) => match.handle?.hydrate);
+
```
Comment (chrisvxd) [inline recipes/remix/app/root.tsx:24]: I'm a bit 50/50 about whether we should include this. It's great, but also bloats the recipe for users that want JS.

Let me have a think on it before we remove it.

File: recipes/remix/app/routes/$puckPath_.edit.tsx [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export { default } from "./edit";
+// I think a bug in remix means loader needs to be explicitly exported here
+export { action, loader } from "./edit";
+// For meta and links etc.
+export * from "./edit";
```
Comment (chrisvxd) [inline recipes/remix/app/routes/$puckPath_.edit.tsx:5]: Smart!

File: recipes/remix/app/routes/edit.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+import puckConfig from "../../puck.config";
+import { getPage, setPage } from "~/models/page.server";
+
+export const action = async ({ params, request }: ActionFunctionArgs) => {
+  const puckPath = params.puckPath || "/";
+  const formData = await request.formData();
+  const puckData = formData.get("puckData");
+
+  // TODO This validation is basic. You should add more here.
```
Comment (chrisvxd) [inline recipes/remix/app/routes/edit.tsx:21]: I think you can remove this comment

File: recipes/remix/app/routes/edit.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  invariant(puckData, "Missing data");
+  invariant(typeof puckData === "string", "Missing data");
```
Comment (chrisvxd) [inline recipes/remix/app/routes/edit.tsx:23]:
```suggestion
  invariant(typeof puckData === "string", "Invalid data");
```

File: recipes/remix/app/routes/edit.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import { Puck, type Data } from "@measured/puck";
+import styles from "@measured/puck/dist/index.css";
```
Comment (chrisvxd) [inline recipes/remix/app/routes/edit.tsx:2]: New format for this now!

```suggestion
import styles from "@measured/puck/puck.css";
```

File: recipes/remix/app/routes/edit.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  /**
+   * TypeStript error
+   * Type 'Config<Props>' is not assignable to type 'Config'.
+   * @see https://github.com/measuredco/puck/issues/185
+   */
+
+  /**
+   * On page load I see the following warning in the browser console. I don't know it it's caused by to Puck or Remix.
+   * Warning: Defining props on `root` is deprecated. Please use `root.props`. This will be a breaking change in a future release. index.js:34421:14
+   */
```
Comment (chrisvxd) [inline recipes/remix/app/routes/edit.tsx:62]: There was a bug in the database.json you used. Try copying the latest database.json from the `next` recipe in canary.

File: recipes/remix/app/routes/_index.tsx [hunk 9]
```diff
@@ -1,41 +1,50 @@
...
+      status: 404,
+      statusText: "Not Found",
+    });
+  }
+  // Return the data.
+  return json({ puckData });
+};
+
+export const meta: MetaFunction<typeof loader> = ({ data }) => {
+  const title = data?.puckData?.root?.title || "Untitled page";
```
Comment (chrisvxd) [inline recipes/remix/app/routes/_index.tsx:35]:
```suggestion
  const title = data?.puckData?.root?.title || "Page";
```

How about this? I think "Untitled" looks worse if users forget to set it.

File: recipes/remix/app/routes/_index.tsx [hunk 10]
```diff
@@ -1,41 +1,50 @@
...
-    </div>
-  );
+export default function Page() {
+  const { puckData } = useLoaderData<typeof loader>();
+
+  /**
+   * TypeStript error
+   * Type 'Config<Props>' is not assignable to type 'Config'.
+   * @see https://github.com/measuredco/puck/issues/185
+   */
```
Comment (chrisvxd) [inline recipes/remix/app/routes/_index.tsx:47]: You can do `puckConfig as Config` for now

**Review by chrisvxd — CHANGES_REQUESTED** [review]
Thanks @EarthlingDavey! This works beautifully 🙌 

I've left some comments.

To get this merged, you'll also need to add the recipe to the [create-puck-app templates](https://github.com/measuredco/puck/tree/main/packages/create-puck-app/templates). You don't need to add all the code, just any handlebars overrides (which will include the package.json - see the [next example](https://github.com/measuredco/puck/tree/main/packages/create-puck-app/templates/next)).

Would also appreciate it if you could add it to the README, below the next recipe.

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (EarthlingDavey) [reply .devcontainer/devcontainer.json:33]: Of course, no problem.

File: recipes/remix/app/routes/edit.tsx (hunk 8 above)
Comment (EarthlingDavey) [reply recipes/remix/app/routes/edit.tsx:61]: I can't see a canary branch, I copied from main and the issue is fixed.

**Comment by EarthlingDavey** [comment]
@chrisvxd thanks for the review. I've worked through them and not touched the disable-js code. I'm not sure about that too. Perhaps that should be stripped but added to an example codebase?

**Comment by chrisvxd** [comment]
Thanks @EarthlingDavey! I think let's remove the no JS code. The user can figure that out based on the Remix docs.

File: recipes/remix/remix.config.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+/** @type {import('@remix-run/dev').AppConfig} */
+export default {
+  ignoredRouteFiles: ["**/.*"],
+  // appDirectory: "app",
+  // assetsBuildDirectory: "public/build",
+  // publicPath: "/build/",
+  // serverBuildPath: "build/index.js",
```
Comment (chrisvxd) [inline recipes/remix/remix.config.js:7]: Presume we can probably remove these?

**Review by chrisvxd — APPROVED** [review]
Awesome stuff, thanks @EarthlingDavey. Merging now and will remove the noJS stuff myself in favour of the remix docs.

**Comment by chrisvxd** [comment]
Can't be merged due to conflicts.

**Pull request merged**